connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -763,6 +789,30 @@ private boolean startTask(ConnectorTaskId taskId) {
         );
     }
 
+    private Callable<Void> getTaskCallable(final ConnectorTaskId taskId, final TaskStatus.State targetState) {
+        return new Callable<Void>() {
+            @Override
+            public Void call() throws Exception {
+                switch (targetState) {
+                    case RUNNING:
+                        try {
+                            startTask(taskId);
```

Comment:
@hachikuji Actually, these semantics are one of the things I'm least happy with about the API design for tasks but which wasn't easily addressed. SourceTasks have the luxury of more asynchronous shutdown, although they do have specific requirements re: having interrupted any outstanding poll() requests (see the SourceTask javadocs). This allows the framework to ensure proper semantics (at least once, exactly once in absence of failures). SinkTasks may have offset commit affected by the stop(), so they need to be more careful to complete shutdown before returning. If they don't, they might not properly guarantee they have fully committed offsets before rebalance (e.g. in the HDFS sink connector you need to fully stop and be sure you've hsynced to HDFS).